### PR TITLE
Peter/customer settings

### PR DESCRIPTION
**Pull request by NerevarineRule**
## Proposed changes

The purpose of this PR is to enable XCODE 11.4 recommended settings by default as per the customer request
(https://github.com/AzureAD/microsoft-authentication-library-for-objc/issues/961) 

Specifically, this PR is aimed at enabling the following three settings by default.

CLANG_ANALYZER_LOCALIZABILITY_NONLOCALIZED = YES; 
CLANG_WARN_DEPRECATED_OBJC_IMPLEMENTATIONS = YES; 
CLANG_WARN_OBJC_IMPLICIT_RETAIN_SELF = YES;

Enabling above settings also require code changes, most of which stem from enabling "CLANG_WARN_OBJC_IMPLICIT_RETAIN_SELF".
Primarily, above setting requires adding self. prefix or self->_ to all properties & ivars being used within block statements.

## Type of change

- [ ] Feature work
- [x] Bug fix
- [ ] Documentation
- [ ] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

## Risk

- [x] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [ ] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [ ] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

### CODE REVIEW

File: IdentityCore/tests/MSIDLastRequestTelemetryTests.m [hunk 1]
```diff
@@ -217,6 +217,8 @@ - (void)testSaveToDisk_whenMultipleSavesThenReset_shouldOverwriteAndRestoreToSam
     
     XCTAssertEqualObjects([restoredTelemetryObject telemetryString], [telemetryObject telemetryString]);
     
+    //[[MSIDLastRequestTelemetry sharedInstance] setValue:@0 forKey:@"silentSuccessfulCount"];
+    //[[MSIDLastRequestTelemetry sharedInstance] setValue:nil forKey:@"errorsInfo"];
```
Comment (kaisong1990) [inline IdentityCore/tests/MSIDLastRequestTelemetryTests.m:221]: Please confider to remove commented out code

File: changelog.txt [hunk 2]
```diff
@@ -13,6 +13,15 @@
 * Add logger connector which allows to override logger behaviour. (#796)
 * Include redirect uri in body when redeeming refresh token at token endpoint (#815)
 
+Version 1.5.5
```
Comment (kaisong1990) [inline changelog.txt:16]: Is the version expected to be here?

File: IdentityCore/src/network/MSIDHttpRequest.m [hunk 3]
```diff
@@ -127,7 +127,7 @@ - (void)sendWithBlock:(MSIDHttpRequestDidCompleteBlock)completionBlock
               
               MSID_LOG_WITH_CTX(MSIDLogLevelVerbose,self.context, @"Parsed response: %@, error %@, error domain: %@, error code: %ld", _PII_NULLIFY(responseObject), _PII_NULLIFY(error), error.domain, (long)error.code);
               
-              if (responseObject && _shouldCacheResponse)
+              if (responseObject && self->_shouldCacheResponse)
```
Comment (kaisong1990) [inline IdentityCore/src/network/MSIDHttpRequest.m:130]: Would you give a try to remove the underscore in front of ```_shouldCacheResponse```

Comment (NerevarineRule) [reply IdentityCore/src/network/MSIDHttpRequest.m:130]: good catch.

File: changelog.txt (hunk 2 above)
Comment (NerevarineRule) [reply changelog.txt:16]: similar comment addressed in MSAL codebase.

File: IdentityCore/src/telemetry/request_telemetry/MSIDLastRequestTelemetry.h [hunk 4]
```diff
@@ -40,7 +40,7 @@ NS_ASSUME_NONNULL_BEGIN
 
 @property (nonatomic, readonly) NSInteger schemaVersion;
 @property (nonatomic, readonly) NSInteger silentSuccessfulCount;
-@property (nonatomic, nullable, readonly) NSArray<MSIDRequestTelemetryErrorInfo *> *errorsInfo;
+@property (nonatomic, nullable, readonly) NSMutableArray<MSIDRequestTelemetryErrorInfo *> *errorsInfo;
```
Comment (oldalton) [inline IdentityCore/src/telemetry/request_telemetry/MSIDLastRequestTelemetry.h:43]: we shouldn't have mutable array in public header

File: IdentityCore/tests/MSIDLastRequestTelemetryTests.m (hunk 1 above)
Comment (oldalton) [reply IdentityCore/tests/MSIDLastRequestTelemetryTests.m:221]: +1

File: IdentityCore/src/cache/metadata/MSIDMetadataCache.m [hunk 5]
```diff
@@ -148,7 +153,7 @@ - (MSIDAccountMetadataCacheItem *)accountMetadataCacheItemWithKey:(MSIDCacheKey
     }
     
     dispatch_barrier_async(_synchronizationQueue, ^{
-        _memoryCache[key] = item;
+        self.memoryCache[key] = item;
```
Comment (oldalton) [inline IdentityCore/src/cache/metadata/MSIDMetadataCache.m:156]: there's now inconsistency between here and line 132. Do you know why it is different?

File: IdentityCore/src/logger/MSIDLogger.m [hunk 6]
```diff
@@ -204,7 +204,7 @@ - (void)logWithLevel:(MSIDLogLevel)level
             
             if (self.nsLoggingEnabled)
             {
-                NSString *logLevelStr = [self stringForLogLevel:_level];
+                NSString *logLevelStr = [self stringForLogLevel:self.level];
```
Comment (oldalton) [inline IdentityCore/src/logger/MSIDLogger.m:207]: will this have any additional impact? accessor for the "level" would go through logger connector. Not sure what is the correct design here. @antrix1989 FYI.

File: IdentityCore/src/cache/metadata/MSIDMetadataCache.m (hunk 5 above)
Comment (NerevarineRule) [reply IdentityCore/src/cache/metadata/MSIDMetadataCache.m:156]: I just have forgotten to apply self. prefix, and XCODE wasn't throwing the warning.

Comment (oldalton) [reply IdentityCore/src/cache/metadata/MSIDMetadataCache.m:156]: Do you know why?

Comment (NerevarineRule) [reply IdentityCore/src/cache/metadata/MSIDMetadataCache.m:156]: As discussed during the scrum meeting, it appears that dispatch_barrier_sync doesn't enforce the requirement of appending self. or self->_ prefix within a block statement when you enable *CLANG_WARN_OBJC_IMPLICIT_RETAIN_SELF*

**Comment by NerevarineRule** [comment]
@oldalton, we agreed during this morning's scrum the following conventions:
1) self. for properties
2) self->_ ivar
3) but for outlier cases like MSIDLastRequestTelemetry, it's ok & acceptable to do self->_  in front of a property to avoid wrong getter being triggered.

File: IdentityCore/src/logger/MSIDLogger.m (hunk 6 above)
Comment (NerevarineRule) [reply IdentityCore/src/logger/MSIDLogger.m:207]: Based on our discussion with Serhii during the scrum on 8/28/2020, it is safe to keep it as self.level and trigger getter api, as there's no harm in calling it.

**Review by oldalton — APPROVED** [review]

**Pull request merged**